bundles/org.openhab.binding.adorne/src/main/java/org/openhab/binding/adorne/internal/hub/AdorneHubController.java

```diff
@@ -0,0 +1,588 @@
+     * updates
+     */
+    private void msgLoop() {
+        try {
+            JsonObject hubMsg;
+            JsonPrimitive jsonService;
+            String service;
+
+            // Main message loop listening for updates from the hub
+            logger.debug("Starting message loop");
+            while (!shouldStop()) {
+                if (!connect()) {
+                    int sleep = hubReconnectSleep;
+                    logger.debug("Waiting {} seconds before re-attempting to connect.", sleep);
+                    if (hubReconnectSleep < HUB_RECONNECT_SLEEP_MAXIMUM) {
+                        hubReconnectSleep = hubReconnectSleep * 2; // Increase sleep time exponentially
+                    }
+                    restartMsgLoop(sleep);
+                    return;
+                } else {
+                    hubReconnectSleep = HUB_RECONNECT_SLEEP_MINIMUM; // Reset
+                }
+
+                if ((hubMsg = getHubMsg()) == null) {
+                    continue;
+                }
+
+                // Process message based on service type
+                if ((jsonService = hubMsg.getAsJsonPrimitive(HUB_TOKEN_SERVICE)) != null) {
+                    service = jsonService.getAsString();
+                } else {
+                    continue; // Ignore messages that don't have a service specified
+                }
+
+                if (service.equals(HUB_SERVICE_REPORT_ZONE_PROPERTIES)) {
+                    processMsgReportZoneProperties(hubMsg);
+                } else if (service.equals(HUB_SERVICE_ZONE_PROPERTIES_CHANGED)) {
+                    processMsgZonePropertiesChanged(hubMsg);
+                } else if (service.equals(HUB_SERVICE_LIST_ZONE)) {
+                    processMsgListZone(hubMsg);
+                } else if (service.equals(HUB_SERVICE_SYSTEM_INFO)) {
+                    processMsgSystemInfo(hubMsg);
+                }
+            }
+        } catch (RuntimeException e) {
+            logger.warn("Hub controller failed", e);
+        }
+
+        // Shut down
+        disconnect();
+        cancelCommands();
+        hubControllerConnected.cancel(false);
+        logger.info("Exiting hub controller");
+    }
+
+    private boolean shouldStop() {
+        long stopTimestamp = this.stopTimestamp;
+        boolean timeoutExceeded = stopTimestamp > 0 && System.currentTimeMillis() > stopTimestamp;
+        boolean stateCommandsIsEmpty;
+        synchronized (stateCommands) {
+            stateCommandsIsEmpty = stateCommands.isEmpty();
+        }
+        boolean commandsServed = stopWhenCommandsServed && stateCommandsIsEmpty && (zoneCommand == null)
+                && (macAddressCommand == null);
+
+        return (isCancelled() || timeoutExceeded || commandsServed);
+    }
+
+    private boolean isCancelled() {
+        Future<?> hubController = this.hubController;
+        return (hubController == null || hubController.isCancelled());
+    }
+
+    private boolean connect() {
```

Comment:
I use the lock to ensure a clean closing and setting to null to prevent a subsequent write or cancel. Creating a new connection on the other hand is ok as there won't be a competing write or cancel when there is no connection.